### PR TITLE
revise week1 & finish 4.1-4.3

### PR DESCRIPTION
**Pull request by eunice012716**
### Todo
- [x] ch2
- [x] ch3 

抱歉PR了很多次QAQ因為linter一直怪怪的

### CODE REVIEW

File: eunice012716/Week1/ch2/2.5/exercise5.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        x.detach().numpy(),
+        x.grad.detach().numpy(),
+        color="green",
+        label="f'(x)",
+    )
+    plt.title("f(x) and its derivative", fontsize=15)
+    plt.xlabel("X", fontsize=13)
+    plt.ylabel("Y", fontsize=13)
+    plt.legend()
+    plt.show()
```
Comment (jamfly) [inline eunice012716/Week1/ch2/2.5/exercise5.py:31]: 可以寫成 function

File: eunice012716/Week1/ch3/3.2/exercise2.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    currents, voltages = synthetic_data(true_w, true_b, 1000)
+
+    w = torch.normal(0, 0.01, size=(1,), requires_grad=True)
+    b = torch.normal(0, 0.01, size=(1,), requires_grad=True)
+    net = linreg
+    loss = squared_loss
+
+    for epoch in range(1, NUM_EPOCHS + 1):
+        for X, y in data_iter(BATCH_SIZE, currents, voltages):
+            loss_ = loss(net(X, w, b), y)  # Minibatch loss in `X` and `y`
```
Comment (jamfly) [inline eunice012716/Week1/ch3/3.2/exercise2.py:60]: 這邊如果你把 loss 改成 get_loss 原本的 loss 就不用加底線，也比較好讀

File: eunice012716/Week1/ch3/3.2/exercise2.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+if __name__ == "__main__":
+    true_w = torch.Tensor([3.3])
+    true_b = 0
+    currents, voltages = synthetic_data(true_w, true_b, 1000)
+
+    w = torch.normal(0, 0.01, size=(1,), requires_grad=True)
+    b = torch.normal(0, 0.01, size=(1,), requires_grad=True)
+    net = linreg
```
Comment (jamfly) [inline eunice012716/Week1/ch3/3.2/exercise2.py:55]: 改成 `get_linreg` or `build_linreg` 都可以，但簡寫都會讓讀的人需要通靈，所以可以寫成 `get_linear_regression`

File: eunice012716/Week1/ch3/3.3/exercise3.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+
+if __name__ == "__main__":
+    data_iter = load_array((FEATURES, LABELS), BATCH_SIZE)
+    net = nn.Sequential(nn.Linear(2, 1))
+    net[0].weight.data.normal_(0, 0.01)
+    net[0].bias.data.fill_(0)
+    loss = nn.MSELoss()
+    trainer = torch.optim.SGD(net.parameters(), lr=0.03)
+    num_epochs = 3
```
Comment (jamfly) [inline eunice012716/Week1/ch3/3.3/exercise3.py:25]: 這也可以是常數

**Review by jamfly — CHANGES_REQUESTED** [review]
可以將 ch4 以後的東西移掉嗎？然後再開一條 PR 做 ch4 還有 ch5

**Comment by jamfly** [comment]
不過整體沒啥問題，一些小問題修掉就可以 merge 了。
